### DISCUSSION TITLE
Gas efficiency of reverting with a custom error vs. reverting through "require"

### DISCUSSION THREAD

**Discussion by ivanmetrikin** [Cyfrin/foundry-full-course-cu#531] [Q&A]
In Lesson 9 Part 3, it's mentioned that reverting with a custom error is more gas efficient than reverting through a `require` statement: https://youtu.be/Og3_o7kFDRw?t=135

The code to prove it is tested through Remix:
<img width="688" alt="image" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/5781858/5f67e98e-d1ea-464e-b54e-20e2613336fe">

However, if I enter the exact same code into Remix, and change the function name slightly (just by 1 character), I get an opposite result: reverting through `require` is **more gas efficient** than reverting with a custom error:
<img width="677" alt="image" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/5781858/f1a58c7d-7405-42aa-aeac-aeae38a03170">

What is the reason for this discrepancy? Namely, why does the addition of "1" character to one of the functions changes the gas efficiency of BOTH functions so dramatically?

**Comment by robbiesumner** [comment]
That's weird, that shouldn't be right.

Did you try actually calling the function in remix on a local chain and then checking the remix terminal output? - maybe it's just a display bug in remix.

**Comment by Niferu** [comment]
Hello @ivanmetrikin 

More gas efficient is using `if(statement)` in every single case. Do not trust gas displayed on remix. Best way to test that is writing some contract including function with if statement. Deploy it and check gas cost of whole deployment, then change if statement to require statement and deploy it again -> compare gas costs results.

If you want to get more insights I highly recommend using `Hardhat` it has tools to measure gas cost of every single function in your contract even in other currency like USD etc.

**Reply by robbiesumner** [reply]
Or just call functions in remix vm and see terminal output

**Reply by Niferu** [reply]
Yes remix (in console) should show same gas costs as tools like hardhat in most cases, but in my opinion Hardhat is way more trust worthy than remix.

**Comment by theirrationalone** [comment]
Why not to go expensive....📈

one bit: 1
one nibble(4 bits): 1111
one byte(2 nibbles | 8 bits): 11111111

one char: 1 byte | 11111111 bits | 1111 - 1111 (2 nibbles)

word of one character -> In bytecode -> 11111111 (2 nibbles) (8 bits) (1 byte)
word of two character -> In bytecode -> 11111111-11111111 (4 nibbles) (16 bits) (2bytes)
word of three character -> In bytecode -> 11111111-11111111-11111111 (6 nibbles) (24 bits) (3 bytes)
.
.
.
and so on....

and this concept, also potentially affects Function's selector... Patrick also explained that sometimes function two or more function's selectors clash with each other...

for example:

We have a function called "withdraw" and its selector is 0x12ghv3asd456z
and we have another function called "anything" and its selector is also 0x12ghv3asd456z,

come back to our gas issue...

Let's take your function:

"revertWithRequire()" -> selector -> 0x34243sf34fs -> taking(assumption) -> 11 * 8 = 88 bits OR 11 * 2 = 22 nibbles OR 11 bytes.

"revertWithError1()" -> selector -> 0xg45sf34sfdfdf3 -> taking(assumption) -> 14 * 8 = 112 bits OR 14*2 = 28 nibbles OR 14 bytes.

So Now, You can see who's the expensive one.... let me break it a bit more....

bytecode understanding ->

```bytecode
0xfunctionSelector_functionParameters_functionDataOrStatements_andIdontKnowIfSomethingElseComesAfterFunctionDataChunk
```

and Please assume that the,
cost of statements of "revertWithRequire()" = 200bytes,
cost of statements of "revertWithError1()" = 198bytes.


"revertWithRequire()" -> bytecode of this function may be like this -> 0x34243sf34fs_nothing_something. -> 11bytes + 0bytes + 200bytes  = 211bytes
"revertWithError1()" -> bytecode of this function may be like this -> 0x34243sf34fs_nothing_something. -> 14bytes + 0bytes +198bytes = 212bytes

211bytes < 212bytes.

So It's clear now, Which one is Clear now.

Disclaimer: I'm not sure about REMIX's implementation on how REMIX calculates Gas Prices.

**Reply by shikhar229169** [reply]
But in the selector we only take 4 bytes. So, size the selector of every function is same which is bytes4. And maybe it should not be the reason to affect gas.

**Reply by theirrationalone** [reply]
I just took different selector lengths to explain the concept.... 

In reality, block producer charges the gas ⛽

But when we're developing the contract, like as we do... 
We can just estimate gas cost. So that's why i took some examples on above discussion... 
That could also be a reason so why we perform symbolic execution after formal verification. 

Just to understand why revert imp. function costs more than require imp. function.... I went out of context a bit to get an idea how might bytecode generated.

**Reply by Niferu** [reply]
@theirrationalone 

But require is also reverting transactions... Those 2 are working exactly the same and using if instead of require cost less.

**Reply by theirrationalone** [reply]
> @theirrationalone 
> 
> But require is also reverting transactions... Those 2 are working exactly the same and using if instead of require cost less.

Damnnnnnn..... YES,

101% we can do this, but at worst revert will win the race........... 

Ther can be some use-cases where require wins over revert but at worst the answer(also recommended) is revert... Effective and saves testing time consumption... 


You can go for more resources to acquire more knowledge some of them patrick provided at last chapters repositories... Slither is the one i remember...

Go acquire more and correct us if we were wrong anywhere, After!

**Reply by Niferu** [reply]
I was just doing some research and testing on mainnet for live project. And that is outcome, those 2 functions are exactly the same, both reverts transactions and are safe to use, just if is cheaper :) I do not know exact calculations like you showed, just did pure testing and sharing outcome with you guys

**Reply by theirrationalone** [reply]
> I was just doing some research and testing on mainnet for live project. And that is outcome, those 2 functions are exactly the same, both reverts transactions and are safe to use, just if is cheaper :) I do not know exact calculations like you showed, just did pure testing and sharing outcome with you guys 

Those calcs are fake and used only to understand internals of gas ⛽calcs... 

Woww research on mainnet 🚀🚀🚀🚀, crazy, would you like to share reports with us??????

**Reply by Niferu** [reply]
for* mainnet not on* mainnet sorry my mistake. I was using hardhat mostly with it's gas reporting feature and in the way I have already explained. Testing summary costs and single function costs changing only if with require without creating additional function to get most relevant results.
